### PR TITLE
feat: support openshift clusters

### PR DESCRIPTION
**Pull request by saumas**

### CODE REVIEW

File: internal/services/discovery/discovery.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+		for _, cond := range n.Status.Conditions {
+			if cond.Type == v1.NodeReady && cond.Status == v1.ConditionTrue {
+				ready = true
+				break
+			}
+		}
+
+		if !ready {
+			continue
+		}
```
Comment (UndeadRat22) [inline internal/services/discovery/discovery.go:112]: could be some isNodeReady helper

Comment (UndeadRat22) [reply internal/services/discovery/discovery.go:112]: Also, more of a question, why can we only check the region/csp on ready nodes?

File: pkg/log/exporter.go [hunk 2]
```diff
@@ -58,6 +61,7 @@ func (ex *exporter) Fire(entry *logrus.Entry) error {
 	ex.wg.Add(1)
 
 	go func(entry *logrus.Entry) {
+		<-ex.waitForRegistrationCh
```
Comment (viktorasm) [inline pkg/log/exporter.go:64]: this channel to block pushing messages is very barebones. it feels like it's a castai.Client responsibility?

```suggestion
		ex.client.WaitForLogsToBeReady(ctx)
```

File: main.go [hunk 3]
```diff
@@ -55,11 +58,20 @@ func main() {
...
 
-	castailog.SetupLogExporter(remoteLogger, localLog, castaiClient, &loggingConfig)
+	waitForRegistrationCh := make(chan struct{})
+	var registrationChClose sync.Once
+	registrationChCloseBody := func() {
+		close(waitForRegistrationCh)
+	}
+	defer func() {
+		registrationChClose.Do(registrationChCloseBody)
+	}()
```
Comment (viktorasm) [inline main.go:68]: this is too cryptic for "main". see my other comment for moving this responsibility to `castaiClient`

File: main.go [hunk 4]
```diff
@@ -230,7 +249,11 @@ func kubeConfigFromPath(kubepath string) (*rest.Config, error) {
 		return nil, nil
 	}
 
-	data, err := ioutil.ReadFile(kubepath)
+	if _, err := os.Stat(kubepath); os.IsNotExist(err) {
+		return nil, err
+	}
+
+	data, err := os.ReadFile(kubepath)
```
Comment (viktorasm) [inline main.go:256]:
```suggestion
	data, err := os.ReadFile(kubepath)
	if os.IsNotExist(err) {
		return nil, err
	}
	
```

File: internal/services/providers/openshift/openshift.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	for {
+		machines, err := p.dyno.Resource(discovery.OpenshiftMachinesGVR).
+			Namespace(discovery.OpenshiftMachineAPINamespace).
+			List(ctx, metav1.ListOptions{Limit: 100, Continue: next})
+		if err != nil {
+			return nil, fmt.Errorf("listing machines: %w", err)
+		}
+
+		for _, machine := range machines.Items {
+			_, ok, err := unstructured.NestedFieldNoCopy(machine.Object, "spec", "providerSpec", "value", "spotMarketOptions")
```
Comment (UndeadRat22) [inline internal/services/providers/openshift/openshift.go:101]: am I reading this correctly, that if the `spotMarketOptions` field exists, that already means that it's a spot instance?

File: internal/services/discovery/discovery.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	return "", false
+}
+
+func getCSP(n *v1.Node) (string, bool) {
+	providerID := n.Spec.ProviderID
+
+	if strings.HasPrefix(providerID, "gce://") {
+		return "gcp", true
```
Comment (viktorasm) [inline internal/services/discovery/discovery.go:160]: we don't have constants for providers in agent?

File: internal/services/discovery/discovery.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+	var properties []string
+	if csp == nil {
+		properties = append(properties, "csp")
+	}
+	if region == nil {
+		properties = append(properties, "region")
+	}
+
+	return nil, nil, fmt.Errorf("failed discovering properties: %s", strings.Join(properties, ", "))
```
Comment (viktorasm) [inline internal/services/discovery/discovery.go:141]: change csp/region to non-pointers, change checks from `==nil` to `==""`, and here just print values.

```suggestion
	return nil, nil, fmt.Errorf("failed discovering cps and region: csp=%q, region=%q", csp, region)
```

File: Makefile [hunk 8]
```diff
@@ -1,5 +1,5 @@
 build:
-	GOOS=linux go build -ldflags="-s -w" -o bin/castai-agent .
+	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags "-s -w" -o bin/castai-agent-amd64 .
```
Comment (viktorasm) [inline Makefile:2]: does `docker build` work after renaming binary?

File: internal/services/discovery/discovery.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	// annotation. kOps annotates the kube-system namespace with annotations such as this:
+	// * addons.k8s.io/core.addons.k8s.io: '{"version":"1.4.0","channel":"s3://bucket/cluster-name/addons/bootstrap-channel.yaml","manifestHash":"hash"}'
+	// We can retrieve the state store bucket name and the cluster name from the "channel" property of the annotation value.
+	GetKOPSClusterNameAndStateStore(ctx context.Context, log logrus.FieldLogger) (clusterName, stateStore *string, reterr error)
+
+	// GetOpenshiftClusterID discovers the OpenShift cluster ID by reading the cluster ID from the OpenShift cluster version resource.
+	GetOpenshiftClusterID(ctx context.Context) (*string, error)
+
+	// GetOpenshiftClusterName discovers the OpenShift cluster name by reading the cluster name from the OpenShift master machine resource.
+	GetOpenshiftClusterName(ctx context.Context) (*string, error)
```
Comment (viktorasm) [inline internal/services/discovery/discovery.go:37]: string pointers are a bit useless, can we use empty strings where we don't see a distinction between empty string and null pointer?

File: internal/services/discovery/discovery.go (hunk 1 above)
Comment (viktorasm) [reply internal/services/discovery/discovery.go:112]: I guess as agent is running, there will be at least one ready node, so maybe they're safer to use:)

File: internal/services/discovery/openshift_test.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	var masterMachineMap map[string]any
+	r.NoError(yaml.Unmarshal([]byte(masterMachineJSON), &masterMachineMap))
+
+	masterMachine := &unstructured.Unstructured{}
+	masterMachine.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   OpenshiftMachinesGVR.Group,
+		Version: OpenshiftMachinesGVR.Version,
+		Kind:    "Machine",
+	})
+	masterMachine.Object = masterMachineMap
```
Comment (viktorasm) [inline internal/services/discovery/openshift_test.go:79]: is this this string-toYAML-to-generic-structure needed? can we just construct an object here?:

```suggestion
	masterMachine.Object = GenericJson{
	   "apiVersion": "machine.openshift...."
	   ...
	   "metadata": GenericJson{
	     "labels": GenericJson {
	         "machine.openshift.io/cluster-api-cluster": "foo-bar"
	     }
	   }
	}
```

File: internal/services/discovery/openshift_test.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	internalID, err := s.GetOpenshiftClusterID(context.Background())
+
+	r.NoError(err)
+	r.Equal("91d87440-5173-47f0-aca5-65bc0144ad30", *internalID)
+}
+
+func TestServiceImpl_GetOpenshiftClusterName(t *testing.T) {
+	r := require.New(t)
+
+	masterMachineJSON := `
```
Comment (viktorasm) [inline internal/services/discovery/openshift_test.go:55]: YAML

File: internal/services/discovery/openshift_test.go (hunk 10 above)
Comment (viktorasm) [reply internal/services/discovery/openshift_test.go:79]: maybe it's not pretty either, when we don't have the golang struct for the "object"

File: internal/services/discovery/discovery.go (hunk 7 above)
Comment (viktorasm) [reply internal/services/discovery/discovery.go:141]: ah, sorry, just saw this is a moved code. kindly asking to boyscout then:)

File: internal/services/discovery/discovery.go (hunk 1 above)
Comment (viktorasm) [reply internal/services/discovery/discovery.go:112]: seems like it's a decision from earlier code, only moved here

File: internal/services/providers/openshift/openshift.go (hunk 5 above)
Comment (saumas) [reply internal/services/providers/openshift/openshift.go:101]: Yes, that is correct. I will add a comment with the link to docs: https://docs.openshift.com/container-platform/4.10/machine_management/creating_machinesets/creating-machineset-aws.html#machineset-creating-non-guaranteed-instance_creating-machineset-aws

File: internal/services/discovery/openshift.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, fmt.Errorf("listing openshift machines: %w", err)
+	}
+
+	if len(machines.Items) == 0 {
+		return nil, fmt.Errorf("no openshift machines found")
+	}
+
+	clusterName, ok := machines.Items[0].GetLabels()[OpenshiftClusterIDLabel]
+	if !ok {
+		return nil, fmt.Errorf("failed to find openshift cluster id")
```
Comment (viktorasm) [inline internal/services/discovery/openshift.go:69]: "name"

for consistency with previous func, maybe align wording as well
```suggestion
		return nil, fmt.Errorf("openshift cluster name not found")
```

File: internal/services/discovery/openshift.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+		List(ctx, metav1.ListOptions{LabelSelector: masterSelector})
+	if err != nil {
+		return nil, fmt.Errorf("listing openshift machines: %w", err)
+	}
+
+	if len(machines.Items) == 0 {
+		return nil, fmt.Errorf("no openshift machines found")
+	}
+
+	clusterName, ok := machines.Items[0].GetLabels()[OpenshiftClusterIDLabel]
```
Comment (viktorasm) [inline internal/services/discovery/openshift.go:67]: I think this fails if labels are null. k8s does not initialize sub-structures with `Get` methods.

File: internal/services/discovery/openshift_test.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	fakedynamic "k8s.io/client-go/dynamic/fake"
+	fakeclientset "k8s.io/client-go/kubernetes/fake"
+
+	"castai-agent/internal/services/controller/scheme"
+)
+
+func TestServiceImpl_GetOpenshiftClusterID(t *testing.T) {
+	r := require.New(t)
+
+	clusterVersionJSON := `
```
Comment (viktorasm) [inline internal/services/discovery/openshift_test.go:20]: yaml

File: internal/services/discovery/openshift_test.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+	var clusterVersionMap map[string]any
+	r.NoError(yaml.Unmarshal([]byte(clusterVersionJSON), &clusterVersionMap))
+
+	version := &unstructured.Unstructured{}
+	version.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   OpenshiftClusterVersionsGVR.Group,
+		Version: OpenshiftClusterVersionsGVR.Version,
+		Kind:    "ClusterVersion",
+	})
+	version.Object = clusterVersionMap
```
Comment (viktorasm) [inline internal/services/discovery/openshift_test.go:39]: would be nice to have a helper string-to-object, e.g.

```suggestion
	version.Object = toYaml(r, `
apiVersion: config.openshift.io/v1
kind: ClusterVersion
metadata:
  name: version
  uid: dc0570f9-2c46-40b1-a5d0-c7233e82c7b6
spec:
  clusterID: 91d87440-5173-47f0-aca5-65bc0144ad30
`)

	... 	
func toYaml(r *require.Assertions, input YAML) map[string] any {
  	var res map[string]any
	r.NoError(yaml.Unmarshal([]byte(inputYAML), &res))
	return res
}
```

File: internal/services/discovery/openshift_test.go (hunk 10 above)
Comment (viktorasm) [inline internal/services/discovery/openshift_test.go:79]: reuse helper here as well for neatness

File: internal/services/providers/openshift/openshift.go [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+	}
+}
+
+func (p *Provider) RegisterCluster(ctx context.Context, client castai.Client) (*types.ClusterRegistration, error) {
+	clusterID, err := p.discoveryService.GetClusterID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting cluster id: %w", err)
+	}
+
+	var csp, region, clusterName, internalID string
```
Comment (viktorasm) [inline internal/services/providers/openshift/openshift.go:41]:
```suggestion
	var openshiftParams castai.OpenshiftParams
```

File: internal/services/providers/openshift/openshift.go [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+	if cfg := config.Get().OpenShift; cfg != nil {
+		csp, region, clusterName, internalID = cfg.CSP, cfg.Region, cfg.ClusterName, cfg.InternalID
```
Comment (viktorasm) [inline internal/services/providers/openshift/openshift.go:44]:
```suggestion
                 openshiftParams.CSP = cfg.CSP
                 ...
```

File: internal/services/providers/openshift/openshift.go [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+	} else {
+		c, r, err := p.discoveryService.GetCSPAndRegion(ctx)
```
Comment (viktorasm) [inline internal/services/providers/openshift/openshift.go:46]:
```suggestion
                 var err
		openshiftParms.CSP, openshiftParams.Region, err = p.discoveryService.GetCSPAndRegion(ctx)
```

File: internal/services/providers/openshift/openshift.go [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("getting openshift cluster id: %w", err)
+		}
+
+		cn, err := p.discoveryService.GetOpenshiftClusterName(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("getting openshift cluster name: %w", err)
+		}
+
+		csp, region, clusterName, internalID = *c, *r, *cn, *id
```
Comment (viktorasm) [inline internal/services/providers/openshift/openshift.go:61]: .. then skip all the temp variables

File: Makefile (hunk 8 above)
Comment (saumas) [reply Makefile:2]: It was broken. I only made it work for my machine. It's in a better state than it was before but obviously not ideal.

File: internal/services/discovery/discovery.go (hunk 6 above)
Comment (saumas) [reply internal/services/discovery/discovery.go]: Added used CSP constants: `[AWS, GCP]`

File: internal/services/discovery/discovery.go (hunk 7 above)
Comment (saumas) [reply internal/services/discovery/discovery.go]: Redone the return values to be point'less :smirk:

File: internal/services/discovery/openshift.go (hunk 13 above)
Comment (saumas) [reply internal/services/discovery/openshift.go]: Added a null check. This shouldn't happen ever but it won't hurt to check.

File: internal/services/discovery/discovery.go (hunk 1 above)
Comment (saumas) [reply internal/services/discovery/discovery.go]: Node objects can be created freely but status conditions are managed by kubelet. We probably got cought on failing to find the data due to some zombie object in the past. Left the readiness check but moved it to a func.

**Review by viktorasm — APPROVED** [review]

**Pull request merged**